dns/handler_whitelist.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2020 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package dns
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/miekg/dns"
+	"github.com/mysteriumnetwork/node/core/policy"
+	"github.com/mysteriumnetwork/node/firewall"
+	"github.com/rs/zerolog/log"
+)
+
+// WhitelistAnswers creates DNS handler which whitelist resolved queries to firewall.
+func WhitelistAnswers(
+	resolver dns.Handler,
+	trafficBlocker firewall.IncomingTrafficBlocker,
+	policies *policy.Repository,
+) dns.Handler {
+	return &whitelistHandler{
+		resolver:       resolver,
+		trafficBlocker: trafficBlocker,
+		policies:       policies,
+	}
+}
+
+type whitelistHandler struct {
+	resolver       dns.Handler
+	trafficBlocker firewall.IncomingTrafficBlocker
+	policies       *policy.Repository
+}
+
+func (wh *whitelistHandler) ServeDNS(writer dns.ResponseWriter, req *dns.Msg) {
+	resolverWriter := &recordingWriter{writer: writer}
+	wh.resolver.ServeDNS(resolverWriter, req)
+	resp := resolverWriter.responseMsg
```

Comment:
You should check the result of DNS resolution. Is there a point in whitelisting if DNS lookup fails, i.e. `resp.Rcode != dns.RcodeSuccess`?

Reply:
DNS resolution answers are traversed here, so errors don't have them.